### ISSUE TITLE
Add Metadata to LLM Calls in Logfire

### ISSUE THREAD

**Issue by PierreColombo** [pydantic/logfire#740]
### Description

Implement a feature in Logfire that allows users to attach custom metadata to each LLM API call. This enhancement would enable more detailed tracking, filtering, and analysis of LLM interactions within the Logfire observability platform and better understanding the tokens consumption.

**Comment by mark-addams001881** [comment]
Hello, we tried to solve the issue.

This is what we did:

Modify the OpenAI integration to allow users to attach custom metadata to each LLM API call. This includes updating the `get_endpoint_config` function to accept and process custom metadata, and modifying the `on_response` function to include this metadata in the span attributes.

You can review changes in this commit: https://github.com/mark-addams001881/pydantic-logfire-740/commit/ba3429362a7a5125496aef3687c9f7e5ad5fd324.


> [!CAUTION]
> Disclaimer: The concept of solution was created by AI and you should never copy paste this code before you check the correctness of generated code. Solution might not be complete, you should use this code as an inspiration only.

---

Latta AI seeks to solve problems in open source projects as part of its mission to support developers around the world. Learn more about our mission at https://latta.ai/ourmission . If you no longer want Latta AI to attempt solving issues on your repository, you can block this account.

**Comment by antoni0z** [comment]
Usign spans allows you to do this, like this:

with logfire.span("ai_flow_1", prompt_id = 5, ... Here you can add as much fields as you need ):
  llm.call() # pseodocode of whatever llm you use. 

And llms in logfire right now are traced by provider (openai, anthropic), I agree that it would be great if they gone deeper into the llm specific filtering apart from doing it from spans, but not clear how without disrupting the openai, anthropic apis. 

Would be great if it can be done generally for all llm type spans.

**Comment by alexmojaki** [comment]
> it would be great if they gone deeper into the llm specific filtering apart from doing it from spans

@antoni0z can you please clarify this?

**Comment by antoni0z** [comment]
> > it would be great if they gone deeper into the llm specific filtering apart from doing it from spans
> 
> @antoni0z can you please clarify this?

Yes, I think that having special attributes for llms is needed.

Because right now, this is being done specific for openai traces and anthropic traces.

So if I want to do a dashboard for the tokens consumed is difficult to do It on a general way on the llms because I have to filter the anthropic way and the openai way.

If for each llm call some attributes are added, like tokens spent, function calls used, and other llm specific attributes this will help with better general llm filtering.

And also will help libraries like lanchain to be able to integrate better with logfire, if it follows this tracing abstractions.

Its just an idea so take it with a grain of salt, what do you think? Does logfire have some of this in mind to become even more helpful in the llms specific context?

**Comment by samuelcolvin** [comment]
Yup, we should use the new semantics Otel attributes for GenAI, we're intending to work on this soon, but it's a big piece of work.

**Comment by antoni0z** [comment]
Amazing to hear that

**Comment by bhaveshpatel640** [comment]
Thanks, looking forward!

**Comment by alexmojaki** [comment]
For attaching custom metadata directly to arbitrary spans that you don't control or have access to, see https://logfire.pydantic.dev/docs/reference/advanced/baggage/
